src/parser/on_demand/parse_to_elements.ts

```diff
@@ -0,0 +1,131 @@
+import { BSONOffsetError } from '../../error';
+import { NumberUtils } from '../../utils/number_utils';
+
+/**
+ * @public
+ * @experimental
+ */
+export type BSONElement = [
+  type: number,
+  nameOffset: number,
+  nameLength: number,
+  offset: number,
+  length: number
+];
+
+/**
+ * Searches for null terminator.
+ * **Does not** bounds check since this should **ONLY** be used within parseToElements which has asserted that `bytes` ends with a `0x00`.
+ * So this will at most iterate to the document's terminator and error if that is the offset reached.
+ */
+function findNull(bytes: Uint8Array, offset: number): number {
+  let nullTerminatorOffset = offset;
+
+  for (; bytes[nullTerminatorOffset] !== 0x00; nullTerminatorOffset++);
+
+  if (nullTerminatorOffset === bytes.length - 1) {
+    throw new BSONOffsetError('Null terminator not found', offset);
+  }
+
+  return nullTerminatorOffset;
+}
+
+/**
+ * @public
+ * @experimental
+ */
+export function parseToElements(bytes: Uint8Array, startOffset = 0): Iterable<BSONElement> {
+  if (bytes.length < 5) {
+    throw new BSONOffsetError(
+      `Input must be at least 5 bytes, got ${bytes.length} bytes`,
+      startOffset
+    );
+  }
+
+  const documentSize = NumberUtils.getSize(bytes, startOffset);
+
+  if (documentSize > bytes.length - startOffset) {
+    throw new BSONOffsetError(
+      `Parsed documentSize (${documentSize} bytes) does not match input length (${bytes.length} bytes)`,
+      startOffset
+    );
+  }
+
+  if (bytes[startOffset + documentSize - 1] !== 0x00) {
+    throw new BSONOffsetError('BSON documents must end in 0x00', startOffset + documentSize);
+  }
+
+  const elements: BSONElement[] = [];
+  let offset = startOffset + 4;
+
+  while (offset <= documentSize + startOffset) {
+    const type = bytes[offset];
+    offset += 1;
+
+    if (type === 0) {
+      if (offset - startOffset !== documentSize) {
+        throw new BSONOffsetError(`Invalid 0x00 type byte`, offset);
+      }
+      break;
+    }
+
+    const nameOffset = offset;
+    const nameLength = findNull(bytes, offset) - nameOffset;
+    offset += nameLength + 1;
+
+    let length: number;
+
+    if (type === 1 || type === 18 || type === 9 || type === 17) {
```

Comment:
These values can _never_ change, the philosophy in my head here is that even though we call them constants, the thinking behind an all-uppercase variable used everywhere is that you can change it and have it be reflected everywhere. I do feel as though the code is "self-documented" because I put the comment and comments are code too 💬

Reply:
I'm with Warren on this - `type === 1` is certainly more opaque than `type === BSONTYPES.DOUBLE` and makes the logic here much more digestible, when you're not intimately familiar with the BSON type identifiers

Reply:
While I have my philosophical (and v. minor practical) reasons for wanting to leave in the literal numbers, I think I found a compromise, added some readable variables at no cost 
🆓 💃🏻